libcoap/src/oscore.rs

```diff
@@ -0,0 +1,142 @@
+        // recipient_id later.
+        let mut initial_recipient: Option<String> = None;
+        let oscore_conf_str = core::str::from_utf8(oscore_conf_bytes).expect("could not parse config bytes to str");
+        for line in oscore_conf_str.lines() {
+            if line.starts_with("recipient_id") {
+                let parts: Vec<&str> = line.split(",").collect();
+                initial_recipient = Some(parts[2].trim().trim_matches('"').to_string());
+                break;
+            }
+        }
+
+        // Return the valid OscoreConf.
+        Ok(Self {
+            raw_conf: oscore_conf,
+            raw_conf_valid: true,
+            initial_recipient,
+        })
+    }
+
+    /// Return the underlying C representation of the oscore config, if its still marked as valid.
+    /// Fails with an error otherwise.
+    ///
+    /// # Errors
+    /// Will return a [OscoreConfigError] if trying to read the raw_config's C struct on an already
+    /// invalidated OscoreConf. Please make sure to only use the OscoreConf once as connect_oscore
+    /// and oscore_server would free the underlying C struct of this config and mark it as invalid.
+    ///
+    /// WARNING: If you clear this pointer you have to mark the raw_conf as invalid by setting the
+    /// raw_conf_valid to false to prevent a double free()!
+    pub(crate) fn as_mut_raw_conf(&mut self) -> Result<*mut coap_oscore_conf_t, OscoreConfigError> {
+        if self.raw_conf_valid {
+            Ok(self.raw_conf)
+        } else {
+            Err(OscoreConfigError::Invalid)
+        }
+    }
+}
+
+impl Drop for OscoreConf {
+    /// Drop the OscoreConf.
+    /// The config will only be dropped, if the raw_struct hasn't been dropped already.
+    fn drop(&mut self) {
+        // SAFETY: Drop the raw_conf if the raw_struct is still valid and hasn't been dropped by
+        // libcoap already. The raw_conf might be freed and invalidated already if connect_oscore
+        // or oscore_server have been called with it previously, in which case the raw_conf_valid
+        // has been set to false, to prevent a double free here.
+        if self.raw_conf_valid {
+            unsafe {
+                let _ = Box::from_raw(self.raw_conf);
+            }
+        }
+    }
+}
+
+/// OscoreRecipient represents a recipient with an ID, and its underlying C struct.
+#[derive(Debug)]
+pub(crate) struct OscoreRecipient {
+    recipient_id: String,
+    recipient: *mut coap_bin_const_t,
+}
+
+impl OscoreRecipient {
+    /// Returns a new OscoreRecipient with a given ID.
+    pub(crate) fn new(recipient_id: &str) -> OscoreRecipient {
+        // The user only supplies the recipients ID, we will build the recipients C struct here.
+        let recipient = coap_bin_const_t {
+            length: recipient_id.len(),
+            s: recipient_id.as_ptr(),
+        };
+
+        let recipient: *mut coap_bin_const_t = Box::into_raw(Box::new(recipient));
```

Comment:
Pointers created by `Box::into_raw` can not necessarily be freed by libcoap. 
To more more specific, providing a different global allocator to Rust code will cause calls to `free()` a Rust-made pointer in libcoap to fail, see https://doc.rust-lang.org/std/boxed/index.html#memory-layout

Use libcoap's own `coap_new_bin_const()` function instead. That way, it is guaranteed that the created pointer can also be freed by libcoap:
```suggestion
        // The user only supplies the recipients ID, we will build the recipients C struct here.
        // SAFETY: provided pointer and length point to a valid byte string usable by coap_new_bin_const()
        let recipient = unsafe { coap_new_bin_const(recipient_id.as_ptr(), recipient_id.len()) };
```

Reply:
Applied to both coap_bin_const_t and coap_oscore_conf_t types.